File: clients/java/examples/with-payload/src/main/java/health/varta/examples/WithPayload.java
package health.varta.examples;

import health.varta.Status;
import health.varta.Varta;

import java.nio.file.Path;
import java.nio.file.Paths;

/** Pack two metrics into the payload field. */
public final class WithPayload {
    private WithPayload() {}

    public static void main(String[] args) throws InterruptedException {
        Path socket = Paths.get(args.length > 0 ? args[0] : "/run/varta/observer.sock");
        try (Varta agent = Varta.connect(socket)) {
            int queueDepth = 0;
            int lastError = 0;
            while (!Thread.currentThread().isInterrupted()) {
                // High 16 bits = queue depth, low 16 bits = last error code.
                int payload = ((queueDepth & 0xFFFF) << 16) | (lastError & 0xFFFF);
                Status status = lastError > 0 ? Status.DEGRADED : Status.OK;
                agent.beat(status, payload);
                queueDepth = (queueDepth + 1) & 0xFFFF;
                Thread.sleep(500);
            }
        }
    }
}

File: crates/varta-client/examples/with_payload.rs
//! Beat loop that packs queue depth and last error code into the 32-bit payload.
//!
//! Layout: high 16 bits = queue_depth (capped), low 16 bits = last_error_code.
//! The observer carries the payload opaquely; decoding is the agent's concern.
//!
//! VLP v0.2 shrunk `payload` from `u64` to `u32` to make room for the
//! CRC-32C wire-integrity trailer. Agents that need more than 4 bytes of
//! context should externalize the data and reference it from the payload
//! (e.g. as a slot index into a shared ring buffer).
//!
//! Run alongside varta-watch:
//!
//! ```sh
//! varta-watch --socket /tmp/varta.sock --threshold-ms 2000 &
//! cargo run --example with_payload
//! ```

use std::sync::atomic::{AtomicU16, Ordering};

static QUEUE_DEPTH: AtomicU16 = AtomicU16::new(0);
static LAST_ERROR: AtomicU16 = AtomicU16::new(0);

fn main() -> std::io::Result<()> {
    let mut agent = varta_client::Varta::connect("/tmp/varta.sock")?;
    loop {
        let depth = QUEUE_DEPTH.load(Ordering::Relaxed);
        let err = LAST_ERROR.load(Ordering::Relaxed);
        let payload = ((depth as u32) << 16) | (err as u32);
        match agent.beat(varta_client::Status::Ok, payload) {
            varta_client::BeatOutcome::Sent => {}
            varta_client::BeatOutcome::Dropped => {
                eprintln!("varta: beat dropped (observer down or queue full)");
            }
            varta_client::BeatOutcome::Failed(e) => {
                eprintln!("varta: beat failed: {e}");
            }
        }
        std::thread::sleep(std::time::Duration::from_millis(500));
    }
}
